src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -0,0 +1,316 @@
+        }
+
+        public void WriteValue(decimal value)
+        {
+            if (this.isIeee754Compatible)
+            {
+                this.writer.WriteStringValue(value.ToString(CultureInfo.InvariantCulture));
+            }
+            else
+            {
+                this.writer.WriteNumberValue(value);
+            }
+
+            this.FlushIfBufferThresholdReached();
+        }
+
+        public void WriteValue(DateTimeOffset value)
+        {
+            this.writer.WriteStringValue(value);
+            this.FlushIfBufferThresholdReached();
+        }
+
+        public void WriteValue(TimeSpan value)
+        {
+            string stringValue = EdmValueWriter.DurationAsXml(value);
+            this.writer.WriteStringValue(stringValue);
+            this.FlushIfBufferThresholdReached();
+        }
+
+        public void WriteValue(Date value)
+        {
+            this.writer.WriteStringValue(value.ToString());
+            this.FlushIfBufferThresholdReached();
+        }
+
+        public void WriteValue(TimeOfDay value)
+        {
+            this.writer.WriteStringValue(value.ToString());
+        }
+
+        public void WriteValue(byte value)
+        {
+            this.writer.WriteNumberValue(value);
+            this.FlushIfBufferThresholdReached();
+        }
+
+        public void WriteValue(sbyte value)
+        {
+            this.writer.WriteNumberValue(value);
+            this.FlushIfBufferThresholdReached();
+        }
+
+        public void WriteValue(string value)
+        {
+            if (value == null)
+            {
+                this.writer.WriteNullValue();
+            }
+            else
+            {
+                this.writer.WriteStringValue(value);
+            }
+
+            this.FlushIfBufferThresholdReached();
+        }
+
+        public void WriteValue(byte[] value)
+        {
+            if (value == null)
+            {
+                this.writer.WriteNullValue();
+            }
+            else
+            {
+                this.writer.WriteBase64StringValue(value);
+            }
+            
+            this.FlushIfBufferThresholdReached();
+        }
+
+        public void WriteRawValue(string rawValue)
+        {
+            if (rawValue == null)
+            {
+                return;
+            }
+
+            // Consider using Utf8JsonWriter.WriteRawValue() in .NET 6+
+            // see: https://github.com/OData/odata.net/issues/2420
+
+            this.Flush(); // ensure we don't write to the stream while there are still pending data in the buffer
+            this.writeStream.Write(this.utf8Encoding.GetBytes(rawValue));
+            this.FlushIfBufferThresholdReached();
+        }
+
+        private void Dispose(bool disposing)
+        {
+            if (this.disposed)
+            {
+                return;
+            }
+
+            if (disposing)
+            {
+                this.writer.Dispose();

```

Comment:
Given that we're buffering writes, what happens if `Dispose` method is called before flushing any pending writes?

Reply:
If a `Stream` was passed to the `Utf8JsonWriter` constructor, then calling `writer.Dispose()` will write to the stream and flush it.

This is confirmed in the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.utf8jsonwriter.flush?view=net-6.0#remarks):

> In the case of [Stream](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream?view=net-6.0), this writes the data to the stream and flushes it.

Reply:
Should I add a note clarifying that in the comments to avoid confusion?

Reply:
I think in most scenarios invoking `Flush` explicitly before invoking `Dispose` helps give someone peace of mind.